File: Code/C_Code/7_3_ColorfulSoftlight/ColorfulSoftlight.cpp
/**********************************************************************
* Filename    : Softlight.cpp
* Description : Use potentiometer to control LED
* Author      : www.freenove.com
* modification: 2021/1/1
**********************************************************************/
#include <wiringPi.h>
#include <stdio.h>
#include <softPwm.h>
#include <ADCDevice.hpp>

#define ledRedPin 21         //define 3 pins for RGBLED
#define ledGreenPin 22
#define ledBluePin 23

ADCDevice *adc;  // Define an ADC Device class object

int main(void){
    adc = new ADCDevice();
    printf("Program is starting ... \n");

    if(adc->detectI2C(0x48)){     // Detect the ads7830
        delete adc;               // Free previously pointed memory
        adc = new ADS7830(0x48);      // If detected, create an instance of ADS7830.
    }
    else{
        printf("No correct I2C address found, \n"
        "Please use command 'i2cdetect -y 1' to check the I2C address! \n"
        "Program Exit. \n");
        return -1;
    }
    wiringPiSetup();
    softPwmCreate(ledRedPin,0,100);     //creat 3 PMW output pins for RGBLED
    softPwmCreate(ledGreenPin,0,100);
    softPwmCreate(ledBluePin,0,100);
    while(1){
        int val_Red = adc->analogRead(2);  //read analog value of 3 potentiometers
        int val_Green = adc->analogRead(3);
        int val_Blue = adc->analogRead(4);
        softPwmWrite(ledRedPin,100-val_Red*100/255);    //map the read value of potentiometers into PWM value and output it
        softPwmWrite(ledGreenPin,100-val_Green*100/255);
        softPwmWrite(ledBluePin,100-val_Blue*100/255);
        
        //print out the read ADC value
        printf("ADC value val_Red: %d  ,\tval_Green: %d  ,\tval_Blue: %d \n",val_Red,val_Green,val_Blue);
        delay(100);
    }
    return 0;
}

File: Code/Python_GPIOZero_Code/7_3_ColorfulSoftlight/ColorfulSoftlight.py
#!/usr/bin/env python3
#############################################################################
# Filename    : Softlight.py
# Description : Control RGBLED with Potentiometer 
# Author      : www.freenove.com
# modification: 2024/07/29
########################################################################
from gpiozero import RGBLED
import time
from ADCDevice import *

led = RGBLED(red=5, green=6, blue=13, active_high=False) 
adc = ADCDevice(0x48) # Define an ADCDevice class object

def setup():
    global adc
    if(adc.detectI2C(0x48)): # Detect the pcf8591.
        adc = ADS7830(0x48)
    else:
        print("No correct I2C address found, \n"
        "Please use command 'i2cdetect -y 1' to check the I2C address! \n"
        "Program Exit. \n");
        exit(-1)
    
def loop():
    while True:     
        value_Red = adc.analogRead(2)       # read ADC value of 3 potentiometers
        value_Green = adc.analogRead(3)
        value_Blue = adc.analogRead(4)
        led.red   =value_Red/255  # map the read value of potentiometers into PWM value and output it 
        led.green =value_Green/255
        led.blue  =value_Blue/255
        # print read ADC value
        print ('ADC Value value_Red: %d ,\tvlue_Green: %d ,\tvalue_Blue: %d'%(value_Red,value_Green,value_Blue))
        time.sleep(0.01)

def destroy():
    adc.close()
    led.close()
    
if __name__ == '__main__': # Program entrance
    print ('Program is starting ... ')
    try:
        setup()
        loop()
    except KeyboardInterrupt:  # Press ctrl-c to end the program.
        print("Ending program")
    finally:
        destroy()
